chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,85 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T<..>).
+    /// Implemented(!T<..>: A).
+    /// Implemented(!T<..>: B).
+    /// Implemented(!T<..>: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T<..>` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        debug_heading!("to_program_clauses({:?})", self);
+        builder.push_binders(&self.bound, |builder, opaque_ty_bound| {
+            let interner = builder.interner();
+            let substitution = builder.substitution_in_scope();
+            let alias = AliasTy::Opaque(OpaqueTy {
+                opaque_ty_id: self.opaque_ty_id,
+                substitution: substitution.clone(),
+            });
+
+            let alias_placeholder_ty = Ty::new(
+                interner,
+                ApplicationTy {
+                    name: TypeName::OpaqueType(self.opaque_ty_id),
+                    substitution,
+                },
+            );
+
+            // AliasEq(T<..> = HiddenTy) :- Reveal.
+            builder.push_clause(
+                DomainGoal::Holds(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: opaque_ty_bound.hidden_ty.clone(),
+                    }
+                    .cast(interner),
+                ),
+                iter::once(DomainGoal::Reveal(())),
+            );
+
+            // AliasEq(T<..> = !T<..>).
+            builder.push_fact(DomainGoal::Holds(
+                AliasEq {
+                    alias: alias.clone(),
+                    ty: alias_placeholder_ty.clone(),
+                }
+                .cast(interner),
+            ));
+
+            let substitution = Substitution::from1(interner, alias_placeholder_ty.clone());
+            for bound in &opaque_ty_bound.bounds {
+                // Implemented(!T<..>: Bound).
+                let bound_with_placeholder_ty = bound.substitute(interner, &substitution);
+                builder.push_binders(&bound_with_placeholder_ty, |builder, bound| {
+                    builder.push_fact(bound);
+                });
+            }
+
+            for auto_trait_id in builder.db.auto_traits() {
```

Comment:
Since we need to get rid of this method, what I think we should be doing instead is to look and check:

If we are generate clauses for a goal like 

```rust
Implemented(!T: Foo)
```

where `!T` is an alias placeholder *and* `Foo` is an auto-trait, then we generate these clauses, and otherwise we don't.

The `to_program_clauses` trait doesn't work for this, but we can just a free function or another extension trait.

Reply:
Right, so we generate such clauses only when there's such a goal. Somewhere around `push_auto_trait_impls` probably?

Anyways, I'll just remove this part for now, I'll add it to the list for next steps.